xlml/utils/xpk.py

```diff
@@ -37,98 +40,122 @@ def generate_workload_id(benchmark_id: str) -> str:
           f" --command='{run_cmds}' --device-type={accelerator_type}"
           f" --num-slices={num_slices} --docker-image={docker_image}"
+          f" --project={cluster_project} --zone={zone}"
       ),
+      "rc=$?",
+      "rm -rf $TMPDIR",
+      "exit $rc",
   )
 
-  subprocess.run(["bash", "-c", ";".join(cmds)], check=True)
+  res = run(["bash", "-c", ";".join(cmds)], stderr=STDOUT, stdout=PIPE)
+  print(res.stdout.decode())
+  assert res.returncode == 0, f"XPK had non-zero return code: {res.returncode}"
 
 
-# To support local execution using `scripts/local-airflow.sh`, run using
-# task.docker in local environments. Otherwise, task.kubernetes should be used.
-if "XLMLTEST_LOCAL_AIRFLOW" in os.environ:
-  run_workload = task.docker(
-      _run_workload, image="google/cloud-sdk:alpine", auto_remove="force"
-  )
-else:
-  run_workload = task.kubernetes(
-      _run_workload,
-      namespace="composer-user-workloads",
-      image="google/cloud-sdk:alpine",
-      config_file="/home/airflow/composer_kube_config",
-      kubernetes_conn_id="kubernetes_default",
-      container_resources=k8s_models.V1ResourceRequirements(
-          limits={"ephemeral-storage": "10G"},
-      ),
-  )
-
-
-# XPK tests are scheduled by Kueue. Allow up to 20 hours for the workload
-# to finish, since it may wait in the queue.
-@task.sensor(
-    poke_interval=60,
-    timeout=timedelta(hours=20).total_seconds(),
-    mode="reschedule",
-)
-def wait_for_workload_completion(
-    workload_id: str, project_id: str, region: str, cluster_name: str
-) -> bool:
-  """Check the workload status."""
+def _get_core_api_client(
+    project_id: str, region: str, cluster_name: str
+) -> k8s_client.CoreV1Api:
+  """Create a core API client for the given cluster."""
   client = gke.get_authenticated_client(project_id, region, cluster_name)
 
   # Initilize the client
   core_api = k8s_client.CoreV1Api(client)
   logging.info("Successful initilize k8s client from cluster response.")
+  return core_api
 
-  # Get pods for the workload
+
+def _list_workload_pods(
+    core_api: k8s_client.CoreV1Api, workload_id: str
+) -> k8s_client.V1PodList:
+  """List all pods for the given workload."""
   logging.info(f"Getting pods for workload_id: {workload_id}")
   pods = core_api.list_namespaced_pod(
       label_selector=f"jobset.sigs.k8s.io/jobset-name={workload_id}",
       namespace="default",
   )
+  return pods
+
+
+@task.sensor(poke_interval=60, timeout=600, mode="reschedule")
+def wait_for_workload_start(
+    workload_id: str, project_id: str, region: str, cluster_name: str
+) -> bool:
+  """Check if the workload has started."""
+  core_api = _get_core_api_client(project_id, region, cluster_name)
+  pods = _list_workload_pods(core_api, workload_id)
+  print(f"Found {len(pods.items)} pods for workload {workload_id}")
+  return len(pods.items) > 0
+
+
+@task.sensor(poke_interval=60, timeout=600, mode="reschedule")
+def wait_for_workload_completion(
```

Comment:
This looks similar to `stream_logs` in `gke.py`. Can these be combined? https://github.com/GoogleCloudPlatform/ml-auto-solutions/blob/6465233478e1bee0a24a1c9019c623b05220e11a/xlml/utils/gke.py#L71

My implementation is flawed, so I'd like to piggy-back on any improvements you've made :smile:

Reply:
I thought about this, but for XPK the workloads can be long-running (up to 5 hours for convergence tests), so I thought to keep it as a sensor to avoid blocking an Airflow worker on the training job.

We're also generally looking at the workloads after execution completes, so I think the cons of occupying the Airflow workers outweighs the benefit, for XPK at least.

Let me know what you think.